### PR TITLE
[testing workflows] Only re-run failed e2e tests

### LINKED ISSUE #49543: Explore only running failed e2e tests on job re-run

### PR DESCRIPTION
**Pull request by adimoldovan**
### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/49543

Use [Playwright's last failed](https://playwright.dev/docs/running-tests#run-last-failed-tests) option to only run previously failed tests on a job re-run.

After the tests run the `.last-run.json` file is uploaded. Before running tests the file is downloaded (if found) and the content is parsed. If failed tests are found the `--last-failed` flag is appended to the tests run command, running only the failed tests.

If no last run info is found (no previous run, or artifact was deleted) all tests should run.

### How to test the changes in this Pull Request:

Re-run some e2e test jobs. Try with failed tests and also with all passed tests. 

I forced some tests to fail and pass one by one based on the run number:
- [Run 1](https://github.com/woocommerce/woocommerce/actions/runs/10255320812/job/28372056118#step:9:88): all (11) tests ran, 3 tests failed - there was no last run info, running all tests
- [Run 2](https://github.com/woocommerce/woocommerce/actions/runs/10255320812/job/28372675500#step:9:90): 3 tests ran, 2 test failed - running only failed tests from Run 1
- [Run 3](https://github.com/woocommerce/woocommerce/actions/runs/10255320812/job/28372849435#step:9:90): 2 tests ran, 1 test failed - running only failed tests from Run 2
- [Run 4](https://github.com/woocommerce/woocommerce/actions/runs/10255320812/job/28373017678#step:9:92): 1 test ran, no failures - running only failed tests from Run 3
- [Run 5](https://github.com/woocommerce/woocommerce/actions/runs/10255320812/job/28373175898#step:9:89): all (11) tests ran, 3 tests failed - no failed tests from Run 4, so all tests will run.
- [Run 6](https://github.com/woocommerce/woocommerce/actions/runs/10255320812/job/28373390576#step:9:90): 3 tests ran, 3 tests failed - running only failed tests from Run 5
- [Run 8](https://github.com/woocommerce/woocommerce/actions/runs/10255320812/job/28373734024#step:9:45): deleted the last run artifact - all (11) tests ran, 3 tests failed
Run 7 had a start environment failure, had to re-run.

### CODE REVIEW

File: plugins/woocommerce/tests/e2e-pw/playwright.config.js [hunk 1]
```diff
@@ -95,7 +95,7 @@ const config = {
 		{
 			name: 'ui',
 			use: { ...devices[ 'Desktop Chrome' ] },
-			testIgnore: '**/api-tests/**',
+			testMatch: [ '**basic.spec.js', '**/product-*spec.js' ],
```
Comment (adimoldovan) [inline plugins/woocommerce/tests/e2e-pw/playwright.config.js]: Updated this to have faster feedback in CI.
To revert before merging. 
```suggestion
			testIgnore: '**/api-tests/**',
```

File: plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js [hunk 2]
```diff
@@ -3,6 +3,7 @@ const { logIn } = require( '../utils/login' );
 const { admin, customer } = require( '../test-data/data' );
 
 test( 'Load the home page', async ( { page } ) => {
+	expect( 1 ).toBe( 2 );
```
Comment (adimoldovan) [inline plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js:6]: Force a test failure.
To revert before merging.

```suggestion

```

File: plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js [hunk 3]
```diff
@@ -17,6 +20,9 @@ test( 'Load the home page', async ( { page } ) => {
 } );
 
 test( 'Load wp-admin as admin', async ( { page } ) => {
+	expect(
+		parseInt( process.env.GITHUB_RUN_ATTEMPT, 10 ) % 2 === 0
+	).toBeFalsy();
```
Comment (adimoldovan) [inline plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js:25]: Forced failure for odd run attempts. To revert before merging.

```suggestion
```

File: plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js [hunk 4]
```diff
@@ -3,6 +3,7 @@ const { logIn } = require( '../utils/login' );
 const { admin, customer } = require( '../test-data/data' );
 
 test( 'Load the home page', async ( { page } ) => {
+	expect( process.env.GITHUB_RUN_ATTEMPT ).toBe( '1' );
```
Comment (adimoldovan) [inline plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js:6]: Forced failure for even run attempts. To revert before merging.

```suggestion
```

File: plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js [hunk 5]
```diff
@@ -17,6 +18,7 @@ test( 'Load the home page', async ( { page } ) => {
 } );
 
 test( 'Load wp-admin as admin', async ( { page } ) => {
+	expect( process.env.GITHUB_RUN_ATTEMPT ).toBe( '2' );
```
Comment (adimoldovan) [inline plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js:21]: Forced failure. To revert before merging.

```suggestion
```

File: plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js [hunk 6]
```diff
@@ -26,6 +28,7 @@ test( 'Load wp-admin as admin', async ( { page } ) => {
 } );
 
 test( 'Load my account page as customer', async ( { page } ) => {
+	expect( process.env.GITHUB_RUN_ATTEMPT ).toBe( '3' );
```
Comment (adimoldovan) [inline plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js:31]: Forced failure. To revert before merging.

```suggestion
```

File: plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js [hunk 7]
```diff
@@ -3,6 +3,7 @@ const { logIn } = require( '../utils/login' );
 const { admin, customer } = require( '../test-data/data' );
 
 test( 'Load the home page', async ( { page } ) => {
+	expect( process.env.GITHUB_RUN_ATTEMPT ).toBe( '2' );
```
Comment (adimoldovan) [inline plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js]: Forced failure. To revert before merging.

```suggestion
```

File: plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js [hunk 8]
```diff
@@ -17,6 +18,7 @@ test( 'Load the home page', async ( { page } ) => {
 } );
 
 test( 'Load wp-admin as admin', async ( { page } ) => {
+	expect( process.env.GITHUB_RUN_ATTEMPT ).toBe( '3' );
```
Comment (adimoldovan) [inline plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js]: Forced failure. To revert before merging.

```suggestion
```

File: plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js [hunk 9]
```diff
@@ -26,6 +28,7 @@ test( 'Load wp-admin as admin', async ( { page } ) => {
 } );
 
 test( 'Load my account page as customer', async ( { page } ) => {
+	expect( process.env.GITHUB_RUN_ATTEMPT ).toBe( '4' );
```
Comment (adimoldovan) [inline plugins/woocommerce/tests/e2e-pw/tests/basic.spec.js]: Forced failure. To revert before merging.

```suggestion
```

**Review by lanej0 — APPROVED** [review]

**Comment by lanej0** [comment]
Reviewed the runs and everything looks good.  Great use of this, should really speed up the re-runs.

**Pull request merged**